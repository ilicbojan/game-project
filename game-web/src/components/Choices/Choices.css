.choices {
  min-height: 136px;
  position: relative;
}

.choices-buttons {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  gap: 1.2rem;
  padding: 1rem 2rem;
}

.choices button {
  flex: 1;
  padding: 1rem;
  font-size: 1.2rem;
  font-weight: bold;
  color: #fff;
  background-color: #007bff;
  border: none;
  border-radius: 5px;
  cursor: pointer;
}

.choices button:disabled {
  background-color: #ccc;
  cursor: not-allowed;
}

@media screen and (max-width: 900px) {
  .choices-buttons {
    display: grid;
    grid-template-columns: 1fr 1fr 1fr;
  }
}

@media screen and (max-width: 550px) {
  .choices-buttons {
    display: grid;
    grid-template-columns: 1fr 1fr;
    gap: 0.5rem;
    padding: 0;
  }

  .choices button {
    font-size: 1rem;
    padding: 0.8rem;
  }
}
