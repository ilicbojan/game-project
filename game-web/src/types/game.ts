export interface ChoiceItem {
  id: number;
  name: string;
}

export interface PlayRequest {
  player: number;
}

export interface PlayResponse {
  player: number;
  computer: number;
  result: ResultType;
}

export interface GameResult {
  player: string;
  computer: string;
  result: ResultType;
  date: Date;
}

export interface Score {
  win: number;
  lose: number;
  tie: number;
}

export enum ResultType {
  Win = "win",
  Lose = "lose",
  Tie = "tie",
}

export enum ChoiceId {
  Rock = 1,
  Paper = 2,
  Scissors = 3,
  Lizard = 4,
  Spock = 5,
}

export enum ChoiceName {
  Rock = "rock",
  Paper = "paper",
  Scissors = "scissors",
  Lizard = "lizard",
  Spock = "spock",
}
