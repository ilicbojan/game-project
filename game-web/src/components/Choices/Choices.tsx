import React from "react";
import { ChoiceItem } from "../../types/game";
import "./Choices.css";
import LoadingSpinner from "../LoadingSpinner/LoadingSpinner";
import { ChoiceToEmojiMap } from "../../utils.ts/helpers";

interface Props {
  choices: ChoiceItem[];
  onChoiceClick: (id: number) => void;
  onRandomClick: () => void;
  loadingChoices: boolean;
  loadingResult: boolean;
}

const Choices: React.FC<Props> = ({
  choices,
  onChoiceClick,
  onRandomClick,
  loadingChoices,
  loadingResult,
}) => {
  return (
    <div className="choices section">
      <h3>Choose your option</h3>
      {loadingChoices && <LoadingSpinner />}
      {!loadingChoices && choices.length === 0 && <p>No choices available</p>}
      {!loadingChoices && choices.length > 0 && (
        <div className="choices-buttons">
          {choices.map((choice) => (
            <button
              key={choice.id}
              onClick={() => onChoiceClick(choice.id)}
              disabled={loadingResult}
            >
              <div>{choice.name.toUpperCase()}</div>
              <div>{ChoiceToEmojiMap[choice.name]}</div>
            </button>
          ))}
          <button onClick={onRandomClick} disabled={loadingResult}>
            <div>RANDOM</div>
            <div>{ChoiceToEmojiMap["random"]}</div>
          </button>
        </div>
      )}
    </div>
  );
};

export default Choices;
