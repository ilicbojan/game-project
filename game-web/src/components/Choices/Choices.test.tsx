import { render, screen, fireEvent } from "@testing-library/react";
import Choices from "../Choices/Choices";
import { ChoiceItem } from "../../types/game";

describe("Choices component", () => {
  const choicesMock: ChoiceItem[] = [
    { id: 1, name: "rock" },
    { id: 2, name: "paper" },
    { id: 3, name: "scissors" },
  ];

  const onChoiceClickMock = jest.fn();
  const onRandomClickMock = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(
      <Choices
        choices={choicesMock}
        onChoiceClick={onChoiceClickMock}
        onRandomClick={onRandomClickMock}
        loadingChoices={false}
        loadingResult={false}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows loading spinner when loadingChoices is true", () => {
    const { container } = render(
      <Choices
        choices={choicesMock}
        onChoiceClick={onChoiceClickMock}
        onRandomClick={onRandomClickMock}
        loadingChoices={true}
        loadingResult={false}
      />
    );
    const spinner = container.querySelector(".loading-spinner");

    expect(spinner).toBeInTheDocument();
  });

  it("renders all choice buttons", () => {
    render(
      <Choices
        choices={choicesMock}
        onChoiceClick={onChoiceClickMock}
        onRandomClick={onRandomClickMock}
        loadingChoices={false}
        loadingResult={false}
      />
    );

    choicesMock.forEach((choice) => {
      expect(screen.getByText(choice.name.toUpperCase())).toBeInTheDocument();
    });
  });

  it("renders the RANDOM button", () => {
    render(
      <Choices
        choices={choicesMock}
        onChoiceClick={onChoiceClickMock}
        onRandomClick={onRandomClickMock}
        loadingChoices={false}
        loadingResult={false}
      />
    );

    expect(screen.getByText("RANDOM")).toBeInTheDocument();
  });

  it("calls onClick handler with correct ID when a button is clicked", () => {
    render(
      <Choices
        choices={choicesMock}
        onChoiceClick={onChoiceClickMock}
        onRandomClick={onRandomClickMock}
        loadingChoices={false}
        loadingResult={false}
      />
    );

    const button = screen.getByText("ROCK");
    fireEvent.click(button);

    expect(onChoiceClickMock).toHaveBeenCalledTimes(1);
    expect(onChoiceClickMock).toHaveBeenCalledWith(1);
  });

  it("calls onRandomClick handler when RANDOM button is clicked", () => {
    render(
      <Choices
        choices={choicesMock}
        onChoiceClick={onChoiceClickMock}
        onRandomClick={onRandomClickMock}
        loadingChoices={false}
        loadingResult={false}
      />
    );

    const randomButton = screen.getByText("RANDOM");
    fireEvent.click(randomButton);

    expect(onRandomClickMock).toHaveBeenCalledTimes(1);
  });

  it("disables all buttons when loading is true", () => {
    render(
      <Choices
        choices={choicesMock}
        onChoiceClick={onChoiceClickMock}
        onRandomClick={onRandomClickMock}
        loadingChoices={false}
        loadingResult={true}
      />
    );

    const buttons = screen.getAllByRole("button");
    buttons.forEach((button) => {
      expect(button).toBeDisabled();
    });
  });

  it("buttons are enabled when loading is false", () => {
    render(
      <Choices
        choices={choicesMock}
        onChoiceClick={onChoiceClickMock}
        onRandomClick={onRandomClickMock}
        loadingChoices={false}
        loadingResult={false}
      />
    );

    const buttons = screen.getAllByRole("button");
    buttons.forEach((button) => {
      expect(button).toBeEnabled();
    });
  });
});
