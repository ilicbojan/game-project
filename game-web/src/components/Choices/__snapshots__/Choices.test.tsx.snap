// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choices component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="choices section"
  >
    <h3>
      Choose your option
    </h3>
    <div
      class="choices-buttons"
    >
      <button>
        <div>
          ROCK
        </div>
        <div>
          ‚úä
        </div>
      </button>
      <button>
        <div>
          PAPER
        </div>
        <div>
          ‚úã
        </div>
      </button>
      <button>
        <div>
          SCISSORS
        </div>
        <div>
          ‚úåÔ∏è
        </div>
      </button>
      <button>
        <div>
          RANDOM
        </div>
        <div>
          üé≤
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
